array.rb

```diff
@@ -1,4 +1,16 @@
 class Array
+  def each
+    unless block_given?
+      return to_enum(:each) { self.length }
+    end
+    i = 0
+    while i < self.length
```

Comment:
I'm not seeing a substantial impact benchmarking the difference, but I wonder if it's worth using a `total = self.length` local variable to avoid the repeated `self.length` method call in `while i < self.length`? [TruffleRuby does this](https://github.com/oracle/truffleruby/blob/vm-22.3.0/src/main/ruby/truffleruby/core/array.rb#L363-L375) in the Ruby implementation of `each_index`.

```ruby
total = self.length
while i < total
```

Reply:
I thought about the same thing. But the current `Array#each` seems to allow extending `self` in the middle of a loop, so it'd be an incompatibility.

Reply:
As to benchmarking, the reason why the issue description is still blank is that both JITs are not ready for optimizing this :) I'll put benchmark results after this becomes indeed useful.